# Assignment 4: Inverse Matrix

The main objective of the assignment is to find the inverse matrix by using two different approaches:LU decomposition and with the function $inv$ from library numpy.linalg.

In [1]:
import random as rn
import numpy as np
from numpy import dot
import time as tm
from numpy.linalg import inv, norm,cond

Firstly, a square matrix A was created with 20 rows and columns was created using the random function from library random using random generator $seed$. Due to the fact that each time the value of $seed$ changes, it had to be provided in the code as: 1519046646. 

In [2]:
SEED = int(tm.time())
rn.seed(1519046646)
n=20
a=np.zeros((n,n),float)
for i in range(n):
    for j in range(n):
        a[i,j]=rn.random()
n=20
    #identity matrix with dimensions 20x20    
b=np.zeros((n,n),float)
b=np.identity(n)
    #initializing matrices x, D, L, U as 2-D arrays:
x=np.zeros((n,n),float)    
D=np.zeros((n,n),float)
L=np.zeros((n,n),float)    
U = np.zeros((n,n),float)

Firstly, two matrices L and U were found using LU decomposition with pivoting.
Then, the function was constructed for calculating $X$ for general case: $AX=B$. In our case, X - inverse matrix to be determined, B - identity matrix. 

In [4]:
def inverseLU(A):    
    for i in range(0,n):
        for k in range(i,n):
            sum=0
            for j in range(0,i):
                sum+=(L[i][j]*U[j][k])
            U[i][k]=A[i][k]-sum
        for k in range(i,n):
            if(i==k):
                L[i][i]=1
            else:
                sum=0
                for j in range(0,i):
                    sum+=L[k][j]*U[j][i]
                L[k][i]=(A[k][i]-sum)/U[i][i]
    #pivoting
    for k in range(i+1, n): 
        if(abs(U[k][i]) > maximum):
            maximum = abs(U[k][i]) 
            m = k        
        for k in range(i, n): 
            temp=U[m][k]
            U[m][k]=U[i][k]
            U[i][k]=temp
    # subtracting lines
        for k in range(i+1,n):
            f = -U[k][i]/float(U[i][i])
            L[k][i] = f                   #used to store multiplier
            for j in range(i, n):
                U[k][j] += f*U[i][j] 
        #making zero below
        for j in range(k+1, n):
            U[j][k]=0
                   
    #Forward substitution LD=b        
    for i in range(0,n):
        D[i] = b[i]/float(L[i][i])
        for k in range(0,i,1):
            D[i] -= D[k]*A[i][k]
    #Backward substitution Ux=D
    for i in range(n-1,-1,-1):
        x[i] = D[i]/float(U[i][i])
        for k in range (i-1,-1,-1):
            U[i] -= x[i]*U[i][k]
    return x   #returns inverse matrix

The accuracy of the value was estimated by finding residual: $$R=|I-AX|$$. Where I is identity matrix with 20 rows and columns. X - calculated value of the inverse. 

In [5]:
r=norm(b-a*x)
print("Residual:", r)
print(norm(x))

Residual: 12.4970611166
19.3928920769


Condition number was calculated by the following formula: $$Cond[A]=|A|*|A^{-1}|$$

In [8]:
condnumber=norm(a)*norm(x)
print(condnumber)

218.793930405


The condition number can also be verified by following formula:
$$|\frac{\delta{x}}{x}|\leq*Cond[A]\|\frac{\delta{A}}{A}|$$

In [9]:
%timeit   ainv = inv(a)

28.4 µs ± 1.29 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [10]:
%timeit    inverseLU(a)

5.61 ms ± 118 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


                                                Conclusion:
Finally, the objective of the experiment was achieved. The inverse matrix was found by using LU decomposition with pivoting. However, due to the pivoting, the inverse matrix was different from that which was obtained by built-in function. 
It can be seen that the residual was equal to about 12.49. However, it cannot indicate whether the system is well-conditioned or not. Next, condition number was equal to approximately 218.79. Since, the well-conditioning is a relative term, there are no particular values that identify the limits of ill-conditioning or well-conditioning of the system. It turned out that the system is consistent, if we get the multiplication within the precision $10^{-16}$. 
Considering the computational efficiency of the implementation and the function $inv$, it can be seen that built-in function is much faster than the implementation above. The difference can be explained by very highly optimized implementation. So, the algorithm behind this function is same, while implementation is different. 
